File: ASLBASILanalysis.m
 function ASLBASILanalysis(SUBJECT, locationASLlabelcontrolPLDNIFTI, locationM0, locationMask, outputmap, PLDlist, locationBASILinfo, artoff, spatialoff)
% ClinicalASL toolbox 2023, JCWSiero
PLDstring = sprintf('%.05g,' , SUBJECT.PLDS(PLDlist));
PLDstring = PLDstring(1:end-1);% strip final comma

if nargin >= 7
    if isempty(numel(locationBASILinfo))
        locationBASILinfostring=[];
    else
        locationBASILinfostring=[' --model-options=' locationBASILinfo];
    end
else
    locationBASILinfostring=[];
end

if nargin >= 8
    if strcmp (artoff, 'artoff')
        artoffstring= ' --artoff'; % do not inver arterial aCBV component, needed for aterial transit time aterfact (ATA) mapping - also much faster, use for per dynamic CBF analysis (outlierremoval)
    else
        artoffstring=[]; % default: inver arterial aCBV component
    end
else
    artoffstring=[];
end

if nargin >= 9
    if strcmp (spatialoff, 'spatialoff')
        spatialstring= 'off'; % do not perform spatial regularization - much faster for per dynamic CBF analysis for outlierremoval
    else
        spatialstring= 'on'; % default: perform spatial regularization
    end
else
    spatialstring='on';
end

T1t = num2str(SUBJECT.T1t);
T1b = num2str(SUBJECT.T1b);
tau = num2str(SUBJECT.tau);
TR_M0 = num2str(SUBJECT.TR_M0(1));
alpha = num2str(SUBJECT.alpha);
slicetime = num2str(SUBJECT.slicetime/1000); % in s
tic

% JCWS % using BASIL precorrected Mz ->Mxy for LooK Locker; fixbolus, well-mixed fit, , -aif-leadscale=0.00001 as set in BASIL_options.txt (TIP from MARTIN GRAIG, Acknowledge him in a paper!
% note basil option --t1 is the tissue T1 for the quantification, --t1t is the tissue T1 for correction the M0 calibration when TR < 5s
system(['oxford_asl -i ' locationASLlabelcontrolPLDNIFTI ' -c ' locationM0 ' -m ' locationMask ' -o ' outputmap ' ' locationBASILinfostring artoffstring ' --spatial=' spatialstring ' --bolus=' tau ...
    ' --slicedt=' slicetime ' --t1=' T1t ' --t1b=' T1b ' --t1t=' T1t ' --plds=' PLDstring ' --tr=' TR_M0 ' --alpha=' alpha ' --iaf=ct --ibf=tis --casl --fixbolus --cmethod voxel --cgain 1.00']);

disp('BASIL analysis finished')
toc
end

File: python/clinical_asl_pipeline/asl_basil_analysis.py
"""
ClinicalASL - Clinical Arterial Spin Labeling processing pipeline

ASL BASIL analysis module.

Repository: https://github.com/JSIERO/ClinicalASL

Author: Jeroen Siero
Institution: UMCU (University Medical Center Utrecht), The Netherlands
Contact: j.c.w.siero@umcutrecht.nl

Description:
    Function for running BASIL (Bayesian Inference for Arterial Spin Labeling) analysis using FSL's oxford_asl tool.

License: BSD 3-Clause License
"""

import subprocess
import time
import logging

def asl_basil_analysis(
    subject,
    location_asl_labelcontrol_pld_nifti,
    location_m0,
    location_mask,
    output_map,
    pld_list,
    artoff=None,
    spatialoff=None
):
# This function runs the BASIL (Bayesian Inference for Arterial Spin Labeling) analysis
# using the FSL oxford_asl tool. It prepares the command line arguments based on the
# provided subject parameters and input file locations, then executes the analysis.
# Parameters:
# subject: dict containing subject information including:
#   - 'T1t': T1 tissue relaxation time in seconds
#   - 'T1b': T1 blood relaxation time in seconds
#   - 'tau': bolus duration in seconds
#   - 'TR_M0': list or tuple, first element is TR for M0 scan in seconds
#   - 'alpha': labeling efficiency (unitless)
#   - 'slicetime': slice timing in milliseconds
# location_asl_labelcontrol_pld_nifti: str, path to ASL label/control PLD NIfTI file
# location_m0: str, path to M0 image file
# location_mask: str, path to brain mask file
# output_map: str, path for output directory
# pld_list: list of PLDs (post-labeling delays) in seconds
# artoff: optional, set to "artoff" to disable arterial component modeling
# spatialoff: optional, set to "spatialoff" to disable spatial regularization

    # Generate comma-separated PLD string
    pld_string = ",".join([f"{pld:.5g}" for pld in pld_list])

    # Arterial component off (optional)
    artoff_string = " --artoff" if artoff == "artoff" else ""

    # Spatial regularization (optional)
    spatial_string = "off" if spatialoff == "spatialoff" else "on"

    # Extract parameter values and convert to string
    T1t = str(subject['T1t'])
    T1b = str(subject['T1b'])
    tau = str(subject['tau'])
    TR_M0 = str(subject['TR_M0'][0])
    alpha = str(subject['alpha'])
    slicetime = str(subject['slicetime'] / 1000)  # convert ms to seconds

    # Timing the execution
    start_time = time.time()

    # Build oxford_asl command
    cmd = (
        f"oxford_asl "
        f"-i {location_asl_labelcontrol_pld_nifti} "
        f"-c {location_m0} "
        f"-m {location_mask} "
        f"-o {output_map} "
        f"{artoff_string} "
        f"--spatial={spatial_string} "
        f"--bolus={tau} "
        f"--slicedt={slicetime} "
        f"--t1={T1t} "
        f"--t1b={T1b} "
        f"--t1t={T1t} "
        f"--plds={pld_string} "
        f"--tr={TR_M0} "
        f"--alpha={alpha} "
        f"--iaf=ct "
        f"--ibf=tis "
        f"--casl "
        f"--fixbolus "
        f"--cmethod voxel "
        f"--cgain 1.00"
    )

    # Run command
    logging.info("Running BASIL analysis...")
    subprocess.run(cmd, shell=True, check=True)

    logging.info("BASIL analysis finished")
    elapsed = round(time.time() - start_time, 2)
    logging.info(f"..this took: {elapsed} s")
